keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md

```diff
@@ -0,0 +1,294 @@
+
+Specifically, this would involve injecting custom behavior during metric registration process by wrapping metric definition, instantiation and enrollment.
+
+### Metric Definition Phase
+
+Currently, the metric definition phase looks like this:
+
+```go
+var someMetricDefinition = prometheus.CounterOpts{
+    Name: "some_metric",
+    Help: "some description",
+}
+```
+
+Since we are using the prometheus provided struct, we are constrained to prometheus provided fields. However, using a custom struct affords us the following:
+```go
+var deprecatedMetricDefinition = kubemetrics.CounterOpts{
+    Name: "some_deprecated_metric",
+    Help: "some description",
+    StabilityLevel: kubemetrics.STABLE, // this is also a custom metadata field
+    DeprecatedVersion: "1.15", // this is a custom metadata field
+}
+
+var alphaMetricDefinition = kubemetrics.CounterOpts{
+    Name: "some_alpha_metric",
+    Help: "some description",
+    StabilityLevel: kubemetrics.ALPHA, // this is also a custom metadata field
+	DeprecatedVersion: "1.15", // this can optionally be included on alpha metrics, although there is no change to contractual stability guarantees
+}
+```
+
+### Metric Instantiation Phase
+
+Currently, the metric instantiation phase looks like this:
+
+```go
+var someCounterVecMetric = prometheus.NewCounterVec(
+    someMetricDefinition,
+    []string{"some-label", "other-label"},
+}
+```
+
+Wrapping the prometheus constructors would allow us to take, as inputs, the modified metric definitions defined above, returning a custom kubernetes metric object which contains the metric which would have been instantiated as well as the custom metadata:
+
+```go
+var deprecatedMetric = kubemetrics.NewCounterVec( // this is a wrapped initializer, which takes in our custom metric definitions
+    deprecatedMetricDefinition, // this is our custom wrapped metric definition from above
+    []string{"some-label", "other-label"},
+}
+var alphaMetric = kubemetrics.NewCounterVec{
+    alphaMetricDefinition, // this is also our custom wrapped metric definition from above
+    []string{"some-label", "other-label"},
+}
+```
+
+### Metric Enrollment Phase
+
+Currently, metric enrollment involves calls to a prometheus function which enrolls the metric in a global registry, like so:
+```go
+prometheus.MustRegister(someCounterVecMetric)
+```
+
+Wrapping a prometheus registry with a kubernetes specific one, would allow us to take our custom metrics from our instantiation phase and execute custom logic based on our custom metadata. Our custom registry would hold a reference to a prometheus registry and defer metric enrollment unless preconditions were met:
+
+```go
+import version "k8s.io/apimachinery/pkg/version"
+
+type Registry struct {
+    promregistry *prometheus.Registry
+    KubeVersion version.Info
+}
+
+// inject custom registration behavior into our registry wrapper
+func (r *Registry) MustRegister(metric kubemetrics.Metric) {
```

Comment:
How do we ensure that users don't keep registering metrics against the existing global registry, effectively by-passing this mechanism? (I can only see this happening by intercepting the marshaling process, called "Gather" in the Prometheus client)

Reply:
You could attempt to register metrics against the global registry but it wouldn't do anything since, presumably, we'd be using our custom registry http handler rather than the global Prometheus handler (i.e. there would no longer be a call to `prometheus.Handler()` anywhere in the code).

Reply:
Are you suggesting we write all the marshaling logic ourselves?

Reply:
No, I was referring to using `promhttp.HandlerFor` instead.